habitat-lab/habitat/tasks/rearrange/actions/oracle_nav_action.py

```diff
@@ -261,6 +261,253 @@ def step(self, *args, **kwargs):
+            self.humanoid_controller.reset(
+                self.cur_articulated_agent.base_transformation
+            )
+        return (agent_pos, np.array(look_at_pos))
+
+    def _path_to_point(self, point):
+        """
+        Obtain path to reach the coordinate point. If agent_pos is not given
+        the path starts at the agent base pos, otherwise it starts at the agent_pos
+        value
+        :param point: Vector3 indicating the target point
+        """
+        agent_pos = self.cur_articulated_agent.base_pos
+
+        path = habitat_sim.ShortestPath()
+        path.requested_start = agent_pos
+        path.requested_end = point
+        found_path = self._sim.pathfinder.find_path(path)
+        if not found_path:
+            return [agent_pos, point]
+        return path.points
+
+    def step(self, *args, **kwargs):
+        # mode = 0,1,2
+        # 0 means both target and lookat are given, 1 means only lookat is given, 2 means only target is given
+        self.skill_done = False
+        nav_to_target_coord = kwargs.get(
+            self._action_arg_prefix + "oracle_nav_lookat_action",
+        )
+        nav_position_coord = kwargs.get(
+            self._action_arg_prefix + "oracle_nav_coord_action",
+        )
+        mode = kwargs.get(self._action_arg_prefix + "mode")
+        if np.linalg.norm(nav_to_target_coord) == 0:
+            return {}
+
+        final_nav_targ, obj_targ_pos = self._get_target_for_coord(
+            nav_to_target_coord, nav_position_coord, mode
+        )
+
+        base_T = self.cur_articulated_agent.base_transformation
+        curr_path_points = self._path_to_point(final_nav_targ)
+        robot_pos = np.array(self.cur_articulated_agent.base_pos)
+        if curr_path_points is None:
+            raise Exception
+        else:
+            # Compute distance and angle to target
+            if len(curr_path_points) == 1:
+                curr_path_points += curr_path_points
+            cur_nav_targ = curr_path_points[1]
+            forward = np.array([1.0, 0, 0])
+            robot_forward = np.array(base_T.transform_vector(forward))
+
+            # Compute relative target.
+            rel_targ = cur_nav_targ - robot_pos
+
+            # Compute heading angle (2D calculation)
+            robot_forward = robot_forward[[0, 2]]
+            rel_targ = rel_targ[[0, 2]]
+            rel_pos = (obj_targ_pos - robot_pos)[[0, 2]]
+
+            angle_to_target = get_angle(robot_forward, rel_targ)
+            angle_to_obj = get_angle(robot_forward, rel_pos)
+
+            dist_to_final_nav_targ = np.linalg.norm(
+                (final_nav_targ - robot_pos)[[0, 2]]
+            )
+            at_goal = (
+                dist_to_final_nav_targ < self._config.dist_thresh
+                and angle_to_obj < self._config.turn_thresh
+            ) or dist_to_final_nav_targ < self._config.dist_thresh / 10.0
+
+            if self.motion_type == "base_velocity":
```

Comment:
It may be cleaner to separate the `base_velocity` and `human_joints` cases in separate functions. Ideally, these should also be enums instead of strings.

Reply:
I thought about it too, but it seems like we would need to pass a lot of arguments if we made them into seperate functions. Happy to hear your suggestions